### PR TITLE
Happy + Lex for parsing

### PR DESCRIPTION
**Pull request by MrChico**
This replaces the current BNFC set up with a Lex + Happy setup. 
All the examples now parse, and all but two typecheck and can be output in the IR format. Obviously more to do but I'd like to get this merged before it gets too large to review.

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -1,2 +1,20 @@
 {nixpkgs ? import <nixpkgs> {}, compiler ? "ghc865"}:
-nixpkgs.pkgs.haskell.packages.${compiler}.callPackage (import ./src/default.nix) {}
+let
+  dapptools = builtins.fetchGit {
```
Comment (leonardoalt) [inline default.nix:3]: Why is this needed?

File: src/Lex.x [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+$space = [\ \t\f\v\r]
+
+tokens :-
+
+  -- ($space* \n)+       { mk BREAK }
+  -- $space+             ;
+  $white+                               ;
+
+  -- reserved words
+  behaviour                             { mk BEHAVIOUR }
```
Comment (leonardoalt) [inline src/Lex.x:20]: Can we also have a
`behavior { mk BEHAVIOUR }`
? ;)

File: src/Lex.x [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+  -- builtin functions
+  newAddr                        { mk NEWADDR }
+
+  -- environment vars
+  CALLER                                { mk CALLER }
+  CALLVALUE                             { mk CALLVALUE }
+  ORIGIN                                { mk ORIGIN }
+  THIS                                  { mk THIS }  -- normally called address, but that's taken
+  NONCE                                 { mk NONCE } -- technically not an opcode
```
Comment (leonardoalt) [inline src/Lex.x:63]: Maybe also add block info?

File: src/Lex.x [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  "="                                   { mk EQ }
+  ">"                                   { mk GT }
+  "<"                                   { mk LT }
+  ":"                                   { mk COLON }
+  "+"                                   { mk PLUS }
+  "-"                                   { mk MINUS }
+  "*"                                   { mk STAR }
+  "/"                                   { mk SLASH }
+  "%"                                   { mk MOD }
+  "^"                                   { mk CARROT }
```
Comment (leonardoalt) [inline src/Lex.x:88]:
```suggestion
  "^"                                   { mk CARET }
```
Unsure if meme

File: src/Parse.y [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+%monad { Err } { (>>=) } { return }
+%tokentype { Lexeme }
+%error { parseError }
+
+%token
+
+  eof                         { L EOF _ }
+
+  -- reserved words
+  'behaviour'                 { L BEHAVIOUR _ }
```
Comment (leonardoalt) [inline src/Parse.y:20]: Same as above, can we add `behavior`?

File: src/test/a.act [hunk 6]
```diff
@@ -0,0 +1,12 @@
+behaviour add of SafeAdd
```
Comment (leonardoalt) [inline src/test/a.act:1]: I think it's better if we make the filenames more clear. Maybe `add.act`?

File: default.nix (hunk 1 above)
Comment (MrChico) [reply default.nix:3]: The nix packages for `hevm` and `sbv` are broken in `nixpkgs` so we need to manually override them right now

File: src/test/a.act (hunk 6 above)
Comment (MrChico) [reply src/test/a.act:1]: sorry, this file shouldn't be here. SafeAdd is already in `tests/safemath`. Will remove this

**Comment by MrChico** [comment]
Thanks @leonardoalt , your comments should now be addressed

File: src/Lex.x [hunk 7]
```diff
@@ -59,6 +60,13 @@ tokens :-
   CALLER                                { mk CALLER }
   CALLVALUE                             { mk CALLVALUE }
   ORIGIN                                { mk ORIGIN }
+  BLOCKHASH                             { mk BLOCKHASH }
+  NUMBER                                { mk NUMBER }
```
Comment (leonardoalt) [inline src/Lex.x:64]: `BLOCKNUMBER` might be less confusing

File: default.nix (hunk 1 above)
Comment (leonardoalt) [reply default.nix:3]: Ok!

File: src/Lex.x (hunk 7 above)
Comment (MrChico) [reply src/Lex.x:64]: yeah, I suppose we can be a bit foreboding of that eip..

**Comment by MrChico** [comment]
@leonardoalt good to merge?
